src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,173 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Out of bounds");
+        }
+
+        MyNode myNode = new MyNode(value);
+
+        if (index == size) {
+
+            add(value);
+        } else if (index == 0) {
+            myNode.next = head;
+            head.prev = myNode;
+            head = myNode;
+            size++;
+        } else {
+
+            MyNode current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            myNode.next = current;
+            myNode.prev = current.prev;
+            current.prev.next = myNode;
+            current.prev = myNode;
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Out of bounds");
+        }
+
+        MyNode current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        MyNode current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = (T) current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        MyNode current = head;
+
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+
+        if (current.prev == null) {
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+        }
+
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        }
+
+        size--;
+        return (T) current.value;
+
     }
 
     @Override
     public boolean remove(T object) {
+
+        MyNode current = head;
+        while (current != null) {
+
+            if (Objects.equals(current.value, object)) {
+
+                if (current.prev == null) {
+                    head = current.next;
+                } else {
+                    current.prev.next = current.next;
+                }
+
+                if (current.next == null) {
+                    tail = current.prev;
+                } else {
+                    current.next.prev = current.prev;
+                }
+
+                size--;
+                return true;
```

Comment:
The unlinking logic is repeated here as well. Using a separate `unlink(MyNode node)` method would make the code cleaner and more maintainable.